### PR TITLE
[Storehouse] 001 - Part 2 Test cases

### PR DESCRIPTION
**Pull request by zhangchiqing**
Accidentally merged https://github.com/onflow/flow-go/pull/8572. 

Now since https://github.com/onflow/flow-go/pull/8569 has been merged, I'm opening this PR again to add test cases for payload node and trie implementation.

## Summary by CodeRabbit

- **Tests**
  - Expanded payloadless trie test coverage, including deterministic root-hash expectations, allocation counting, read/leaf-hash correctness, and unallocation behavior.
  - Added equivalence checks against the regular trie, covering pruning and unsafe inclusion/proof equivalence.
  - Strengthened randomized allocation/unallocation validation to ensure consistent roots, counts, and per-path results.
  - Added JSON dump verification for empty, default/unallocated, and multi-leaf cases.
  - Added interim-node and leaf enumeration hashing assertions, plus payloadless proof/proof-batch behavior checks (construction, equality, and string output).

### CODE REVIEW

File: ledger/complete/payloadless/trie_test.go [hunk 1]
```diff
@@ -1,63 +1,971 @@
...
-		require.Equal(t, base.RootHash(), updated.RootHash())    // writing empty to a fresh path is a no-op
+		require.True(t, maxDepthTouched <= 256)
+		require.Equal(t, payloadless.EmptyTrieRootHash(), updatedTrieNoPruning.RootHash())
+		require.Equal(t, uint64(0), updatedTrieNoPruning.AllocatedRegCount())
+
+		// Update with removed paths and non-empty values
+		// (old value empty and new value present)
+		updatedTrieNoPruning, maxDepthTouched, err = payloadless.NewTrieWithUpdatedRegisters(updatedTrieNoPruning, paths, values, false)
+		require.NoError(t, err)
+		require.Equal(t, rootHash, updatedTrie.RootHash())
```
Comment (AlexHentschel) [inline ledger/complete/payloadless/trie_test.go]: This assertion re-checks `updatedTrie`, which is unchanged since line 739 — so it is trivially true, and the actual property (re-allocating the removed registers restores the original root hash on the no-pruning trie) is unverified.

```suggestion
		require.Equal(t, rootHash, updatedTrieNoPruning.RootHash())
```

I verified locally that the test passes with the corrected assertion. Note the mtrie original has the identical bug (`ledger/complete/mtrie/trie/trie_test.go` line 1044) — candidate for a small follow-up fix there.

File: ledger/complete/payloadless/trie_test.go [hunk 2]
```diff
@@ -1,63 +1,971 @@
...
+	}
+	for i := index; i < len(paths); i++ {
+		assert.Equal(t, bitutils.ReadBit(paths[i][:], randomIndex), 1)
+	}
+
+	// check the multi-set didn't change
+	sort.Slice(paths, func(i, j int) bool {
+		return bytes.Compare(paths[i][:], paths[j][:]) < 0
+	})
+	for i := index; i < len(paths); i++ {
```
Comment (AlexHentschel) [inline ledger/complete/payloadless/trie_test.go]: The loop starts at `index`, so the first partition `paths[0:index]` is never compared against `sortedPaths` — corruption of the zero-bit partition would go undetected. Both slices are fully sorted at this point, so the check can simply cover the whole range:

```suggestion
	for i := 0; i < len(paths); i++ {
```

Same issue in the mtrie original (`ledger/complete/mtrie/trie/trie_test.go` line 413).

File: ledger/complete/payloadless/trie_test.go [hunk 3]
```diff
@@ -1,63 +1,971 @@
...
+	return dedupedPaths, dedupedValues
+}
+
+func TestSplitByPath(t *testing.T) {
+	rand := unittest.GetPRG(t)
+
+	const pathsNumber = 100
+	const redundantPaths = 10
+	const pathsSize = 32
+	randomIndex := rand.Intn(pathsSize)
```
Comment (AlexHentschel) [inline ledger/complete/payloadless/trie_test.go]: `pathsSize = 32` is the path length in **bytes**, but it bounds the **bit** index passed to `SplitPaths`. The test therefore only ever splits on bit indices 0–31 (the first 4 bytes); the remaining 224 bit positions — including the byte-boundary cases inside `bitutils.ReadBit` — are never exercised.

```suggestion
	randomIndex := rand.Intn(pathsSize * 8)
```

(Inherited from the mtrie original.)

File: ledger/complete/payloadless/trie_test.go [hunk 4]
```diff
@@ -1,63 +1,971 @@
...
+func Test_TrieWithMiddleRegister(t *testing.T) {
+	// Make new Trie (independently of MForest):
+	emptyTrie := payloadless.NewEmptyMTrie()
+
+	path := testutils.PathByUint16LeftPadded(56809)
+	value := payloadValue(testutils.LightPayload(12346, 59656))
+	leftPopulatedTrie, maxDepthTouched, err := payloadless.NewTrieWithUpdatedRegisters(emptyTrie, []ledger.Path{path}, [][]byte{value}, true)
+	require.Equal(t, uint16(0), maxDepthTouched)
+	require.Equal(t, uint64(1), leftPopulatedTrie.AllocatedRegCount())
+	require.NoError(t, err)
```
Comment (AlexHentschel) [inline ledger/complete/payloadless/trie_test.go]: Nit: `require.NoError(t, err)` should precede the assertions on the other return values — if the update errors, the `maxDepthTouched` / reg-count failures mask the root cause. Same ordering at lines 220–222, 227–229, and 435–438.

File: ledger/complete/payloadless/trie_test.go [hunk 5]
```diff
@@ -1,63 +1,971 @@
...
+
+	// we then write an additional 117 registers
+	paths2, values2 := deduplicateWrites(sampleRandomRegisterWrites(rng, 117))
+	updatedTrie, maxDepthTouched, err = payloadless.NewTrieWithUpdatedRegisters(updatedTrie, paths2, values2, true)
+	require.Equal(t, uint16(254), maxDepthTouched)
+	require.Equal(t, uint64(len(values1)+len(values2)), updatedTrie.AllocatedRegCount())
+	require.NoError(t, err)
+
+	// and now we override the first 99 registers with default values, i.e. unallocate them
+	emptyValues0 := make([][]byte, len(values1))
```
Comment (AlexHentschel) [inline ledger/complete/payloadless/trie_test.go:226]: All unallocation tests use `nil` values. The spec's boundary condition is `len(value) == 0` (see `NewLeaf`), which the empty non-nil slice `[]byte{}` also satisfies — but that representation is never tested anywhere in the suite. Suggest setting a few of these entries to `[]byte{}` so both representations of "unallocated" are pinned as equivalent.

File: ledger/complete/payloadless/equivalence_test.go [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+// TestEquivalence_UnsafeProofs verifies that the proof interim hashes and
+// structure (Flags, Steps, Inclusion) match between the two implementations.
+// The payload-vs-leafHash field differs by design.
+func TestEquivalence_UnsafeProofs(t *testing.T) {
+	rng := &LinearCongruentialGenerator{seed: 0}
+	paths, values := deduplicateWrites(sampleRandomRegisterWrites(rng, 300))
+
+	m, pl := applyToBoth(t,
+		trie.NewEmptyMTrie(), payloadless.NewEmptyMTrie(),
+		paths, values, true,
```
Comment (AlexHentschel) [inline ledger/complete/payloadless/equivalence_test.go:266]: This test runs with `prune=true` only, so the trie contains no explicit default leaves. Though, proof generation is the reason `prune=false` exists (specifically, as a shortcut for missing non-inclusion proofs). The core mechanism — a query path hitting an *explicitly represented unallocated register*, which `prove()` reports as `Inclusion=true` with `LeafHash=nil` (payloadless) / empty payload (mtrie) — is never exercised, so that encoding is currently unverified by any test.

Suggest adding a `prune=false` variant: allocate registers, unallocate a subset without pruning, then query (a) still-allocated paths, (b) the explicitly-unallocated paths, (c) never-allocated paths. Note the inclusion branch below (`require.NotNil(t, plp.LeafHash)`) then needs to distinguish the default-leaf case: when the mtrie payload is empty, the expected payloadless encoding is `Inclusion=true, LeafHash=nil`.

File: ledger/complete/payloadless/equivalence_test.go [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+			require.NotNil(t, plp.LeafHash)
+			expected := hash.HashLeaf(hash.Hash(mp.Path), []byte(mp.Payload.Value()))
+			require.Equal(t, expected, *plp.LeafHash)
+		}
+	}
+}
+
+// TestEquivalence_RandomWalk runs random allocations, updates, and
+// unallocations on both implementations and checks the root hash, register
+// count, and per-path reads agree after every step.
```
Comment (AlexHentschel) [inline ledger/complete/payloadless/equivalence_test.go]: Doc/code mismatch: the comment promises "per-path reads agree after every step", but the loop only checks root hash and register count. I'd recommend expanding the test to explicitly perform reads (rather than trimming the comment)


Specifically: `UnsafeRead` over the live set plus a few dead paths per step is cheap and would make this the strongest test in the file (it would also cover reads of registers that were first unallocated and later re-allocated across steps, a state-transition sequence no other test hits).

File: ledger/complete/payloadless/node_test.go [hunk 8]
```diff
@@ -487,6 +487,78 @@ func Test_Compactify_BothChildrenPopulated(t *testing.T) {
...
+// the subtree (here: three leaves across a small two-level interim structure).
+func Test_AllLeafHashes(t *testing.T) {
+	path := testutils.PathByUint16(1)
+	v := []byte(testutils.LightPayload(2, 3).Value())
+	n1 := NewLeaf(path, v, 0)
+	n2 := NewLeaf(path, v, 0)
+	n3 := NewLeaf(path, v, 1)
+	n4 := NewInterimNode(1, n1, n2)
+	n5 := NewInterimNode(2, n4, n3)
+	require.Equal(t, 3, len(n5.AllLeafHashes()))
```
Comment (AlexHentschel) [inline ledger/complete/payloadless/node_test.go]: Two weaknesses here:

1. Only the *count* is asserted — an implementation returning the wrong hashes (or the same hash three times) would pass. Suggest asserting the returned set equals the three expected leaf hashes.
2. `n1` and `n2` are the *same* register (identical path and value) attached as both children of `n4`, which is not a well-formed trie shape. Distinct registers keep the fixture representative and make the content assertion from (1) meaningful.

Adding a default leaf (unallocated register) into the subtree would also pin the "empty leaves are skipped" clause of the contract, which is currently untested at the node level.

File: ledger/complete/payloadless/json_test.go [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+	path3 := testutils.PathByUint16(3)
+
+	value1 := []byte{1}
+	value2 := []byte{2}
+	value3 := []byte{3}
+
+	paths := []ledger.Path{path1, path2, path3}
+	values := [][]byte{value1, value2, value3}
+
+	tr, _, err := payloadless.NewTrieWithUpdatedRegisters(payloadless.NewEmptyMTrie(), paths, values, true)
```
Comment (AlexHentschel) [inline ledger/complete/payloadless/json_test.go:105]: Both dump tests use pruned tries, so `DumpAsJSON` on an *explicit default leaf* (as occurs in unpruned tries) is un-tested: it currently emits `{"path": ..., "leafHash": null}`. If that is intended, worth a test case (build with `prune=false`, unallocate one register); if unallocated registers should rather be skipped (as `AllLeafHashes` does), that's a production-code question — either way a test would force the decision to be explicit.

File: ledger/complete/payloadless/node_test.go [hunk 10]
```diff
@@ -487,6 +487,78 @@ func Test_Compactify_BothChildrenPopulated(t *testing.T) {
...
+	require.Equal(t, ledger.GetDefaultHashForHeight(9), n.Hash())
+
+	n = NewInterimNode(16, nil, nil)
+	expectedRootHashHex = "6e24e2397f130d9d17bef32b19a77b8f5bcf03fb7e9e75fd89b8a455675d574a"
+	require.Equal(t, expectedRootHashHex, hashToString(n.Hash()))
+	require.Equal(t, ledger.GetDefaultHashForHeight(16), n.Hash())
+}
+
+// Test_InterimNodeWithOneChild verifies that the hash value of an interim node with
+// only one child (left or right) is computed correctly.
```
Comment (AlexHentschel) [inline ledger/complete/payloadless/node_test.go]: A small diagram would make this hash-pinning test self-explanatory and, more importantly, document *why* the two constructions hash differently (the nil sibling contributes `DefaultHashForHeight(h-1)`):

```suggestion
// Test_InterimNodeWithOneChild verifies that the hash value of an interim node with
// only one child (left or right) is computed correctly. The nil sibling contributes
// DefaultHashForHeight(h-1), which is why the left-only and right-only nodes below hash
// differently:
//
//       NewInterimNode(1, c, nil)          NewInterimNode(1, nil, c)
//               (h=1)                              (h=1)
//               /    \                             /    \
//        c=Leaf(56809)  nil=D(0)          nil=D(0)  c=Leaf(56809)
//            (h=0)                                      (h=0)
```

File: ledger/complete/payloadless/node_test.go [hunk 11]
```diff
@@ -487,6 +487,78 @@ func Test_Compactify_BothChildrenPopulated(t *testing.T) {
...
+	expectedRootHashHex := "aa496f68adbbf43197f7e4b6ba1a63a47b9ce19b1587ca9ce587a7f29cad57d5"
+	require.Equal(t, expectedRootHashHex, hashToString(n.Hash()))
+
+	n = NewInterimNode(1, nil, c)
+	expectedRootHashHex = "9845f2c9e9c067ec6efba06ffb7c1be387b2a893ae979b1f6cb091bda1b7e12d"
+	require.Equal(t, expectedRootHashHex, hashToString(n.Hash()))
+}
+
+// Test_InterimNodeWithBothChildren verifies that the hash value of an interim node with
+// both children (left and right) is computed correctly.
```
Comment (AlexHentschel) [inline ledger/complete/payloadless/node_test.go]: A diagram clarifies the shape under test. Worth also noting the construction is synthetic, so a reader doesn't puzzle over why `56809` (MSB=1) sits on the left:

```suggestion
// Test_InterimNodeWithBothChildren verifies that the hash value of an interim node with
// both children (left and right) is computed correctly:
//
//              NewInterimNode(1, leftChild, rightChild)
//                            (h=1)
//                            /   \
//               leftChild=Leaf    rightChild=Leaf
//                 (56809, h=0)       (2, h=0)
//
// Note: the construction is synthetic — the child paths are not consistent with their branch
// positions (NewInterimNode does not validate this); only the resulting node hash (a function
// of the two child hashes) is under test.
```

File: ledger/complete/payloadless/node_test.go [hunk 12]
```diff
@@ -487,6 +487,78 @@ func Test_Compactify_BothChildrenPopulated(t *testing.T) {
...
+	rightValue := []byte(testutils.LightPayload(11, 22).Value())
+	rightChild := NewLeaf(rightPath, rightValue, 0)
+
+	n := NewInterimNode(1, leftChild, rightChild)
+	expectedRootHashHex := "1e4754fb35ec011b6192e205de403c1031d8ce64bd3d1ff8f534a20595af90c3"
+	require.Equal(t, expectedRootHashHex, hashToString(n.Hash()))
+}
+
+// Test_AllLeafHashes verifies that AllLeafHashes collects the leaf hash of every allocated register in
+// the subtree (here: three leaves across a small two-level interim structure).
```
Comment (AlexHentschel) [inline ledger/complete/payloadless/node_test.go]: The two-level shape is the one non-obvious structure among these tests — a diagram helps. This suggestion only touches the doc comment; it composes with my other comment on the fixture (using three distinct registers), which should be applied to the body:

```suggestion
// Test_AllLeafHashes verifies that AllLeafHashes collects the leaf hash of every allocated register in
// the subtree (here: three leaves across a small two-level interim structure). Structure under test
// (apply the fixture fix from my other comment first, so n1/n2/n3 are three DISTINCT registers):
//
//                    n5 (h=2)
//                   /        \
//               n4 (h=1)      n3 (h=1)
//               /     \
//          n1 (h=0)   n2 (h=0)
```

**Review by AlexHentschel — APPROVED** [review]
Looks good. 

* suggested the addition of clarifying Ascii diagrams in a few places
* AI identified some untested side-cases and proposed extension of tests (these were mostly untested already in the prior version of mTrie). The extension should be straight forward for an AI to implement 🤞 .

Overall, nice work consolidating the tests.

**Review by janezpodhostnik — APPROVED** [review]
Looks like Alex already covered everything I had
